quinn/tests/many_connections.rs

```diff
@@ -0,0 +1,200 @@
+                .then(move |_| Ok(()));
+            current_thread::spawn(task);
+
+            Ok(())
+        })
+        .take(expected_messages as u64)
+        .for_each(|_| Ok(()));
+    runtime.spawn(read_incoming_data);
+
+    let client_cfg = configure_connector(&listener_cert);
+
+    for _ in 0..expected_messages {
+        let data = random_data_with_hash(1024 * 1024);
+        let ctx = ctx.clone();
+        let task = unwrap!(endpoint.connect_with(&client_cfg, &listener_addr, "Test"))
+            .map_err(|e| panic!("Connection failed: {}", e))
+            .and_then(move |(conn_driver, conn, _)| {
+                current_thread::spawn(conn_driver.map_err(|_| ()));
+                write_to_peer(&conn, data);
+                ctx.borrow_mut().connections.push(conn);
+                Ok(())
+            });
+        runtime.spawn(task);
+    }
+    drop(endpoint); // we don't need it anymore, this will make EndpointDriver finish.
+
+    unwrap!(runtime.run());
+}
+
+fn read_from_peer(
+    stream: quinn::NewStream,
+    ctx: Rc<RefCell<TestContext>>,
+    conn: quinn::Connection,
+) {
+    let stream = match stream {
+        quinn::NewStream::Bi(_bi) => panic!("Unexpected bidirectional stream here"),
+        quinn::NewStream::Uni(uni) => uni,
+    };
+
+    let task = quinn::read_to_end(stream, 1024 * 1024 * 5)
+        .map_err(|e| panic!("read_to_end() failed: {}", e))
+        .and_then(move |(_stream, data)| {
+            assert!(hash_correct(&data));
+
+            let mut ctx = ctx.borrow_mut();
+            ctx.received_messages += 1;
+            if ctx.received_messages == ctx.expected_messages {
+                // let done_tx = unwrap!(ctx.test_done_tx.take());
+                // unwrap!(done_tx.send(()));
+            }
+
+            conn.close(0, &[]);
+            Ok(())
+        })
+        .then(|_| Ok(()));
+    current_thread::spawn(task);
+}
+
+fn write_to_peer(conn: &quinn::Connection, data: Vec<u8>) {
+    let task = conn
+        .open_uni()
+        .map_err(|e| panic!("Failed to open unidirection stream: {}", e))
+        .and_then(move |o_stream| {
+            tokio::io::write_all(o_stream, data).map_err(|e| panic!("write_all() failed: {}", e))
+        })
+        .and_then(move |(o_stream, _)| {
+            tokio::io::shutdown(o_stream).map_err(|_| ()) // ignore errors, remote peer might already have hung up
+        })
+        .map(|_| ());
+    current_thread::spawn(task);
+}
+
+/// Builds client configuration. Trusts given node certificate.
+fn configure_connector(node_cert: &[u8]) -> quinn::ClientConfig {
+    let mut peer_cfg_builder = quinn::ClientConfigBuilder::new();
+    let their_cert = unwrap!(quinn::Certificate::from_der(&node_cert));
+    unwrap!(peer_cfg_builder.add_certificate_authority(their_cert));
+    let mut peer_cfg = peer_cfg_builder.build();
+    let transport_config = unwrap!(Arc::get_mut(&mut peer_cfg.transport));
+    transport_config.idle_timeout = 30_0000;
```

Comment:
```suggestion
    transport_config.idle_timeout = 30_000;
```
Unless it actually needs 300 seconds.